src/main/groovy/org/pih/warehouse/inventory/ProductInventorySnapshotTaker.groovy

```diff
@@ -0,0 +1,69 @@
+package org.pih.warehouse.inventory
+
+import grails.validation.ValidationException
+import org.springframework.stereotype.Component
+
+import org.pih.warehouse.api.AvailableItem
+import org.pih.warehouse.core.Location
+import org.pih.warehouse.product.Product
+
+/**
+ * Responsible for creating and persisting product inventory transactions. We call this a "snapshot taker" because
+ * product inventory transactions saved via this class are functionally a copy of what the quantity on hand
+ * (determined by product availability) was for each [bin location + lot number] combo of the product at the time.
+ * Importantly, transactions created via this class should NOT result in any quantity changes/adjustments.
+ */
+@Component
+class ProductInventorySnapshotTaker {
+
+    ProductAvailabilityService productAvailabilityService
+    TransactionIdentifierService transactionIdentifierService
+
+    /**
+     * Take a new product inventory "snapshot" transaction based on the current product availability.
+     */
+    Transaction createTransaction(
+            Location facility,
+            Product product,
+            ProductInventorySnapshotSource sourceType,
+            Object source,
+            Date transactionDate=new Date()) {
+
+        TransactionType transactionType = TransactionType.read(sourceType.transactionTypeId)
+
+        Transaction transaction = new Transaction(
+                source: facility,
+                inventory: facility.inventory,
+                transactionDate: transactionDate,
+                transactionType: transactionType,
+        )
+
+        transaction.transactionNumber = transactionIdentifierService.generate(transaction)
+
+        switch (sourceType) {
+            case ProductInventorySnapshotSource.CYCLE_COUNT:
+                transaction.cycleCount = source as CycleCount
```

Comment:
if you cast it to `CycleCount` here, why can't it become a type of the argument, instead of having `Object source`?